source/slang/slang-ir.cpp

```diff
@@ -3300,19 +3301,244 @@ namespace Slang
+                constexpr int maxCount = 4096;
+                if (count->getValue() > maxCount)
+                    break;
+                for (IRIntegerValue i = 0; i < count->getValue(); i++)
+                {
+                    elements.add(element);
+                }
+                return emitMakeArray(type, elements.getCount(), elements.getBuffer());
+            }
+            break;
+        }
+        case kIROp_VectorType:
+        {
+            auto inner = emitDefaultConstruct(as<IRVectorType>(actualType)->getElementType(), fallback);
+            if (!inner)
+                return nullptr;
+            return emitIntrinsicInst(type, kIROp_constructVectorFromScalar, 1, &inner);
+        }
+        case kIROp_MatrixType:
+        {
+            auto inner = emitDefaultConstruct(as<IRMatrixType>(actualType)->getElementType(), fallback);
+            if (!inner)
+                return nullptr;
+            return emitIntrinsicInst(type, kIROp_MakeMatrixFromScalar, 1, &inner);
+        }
+        default:
+            break;
+        }
+        if (fallback)
+        {
+            return emitIntrinsicInst(type, kIROp_DefaultConstruct, 0, nullptr);
+        }
+        return nullptr;
+    }
+
+    static int _getTypeStyleId(IRType* type)
+    {
+        if (auto vectorType = as<IRVectorType>(type))
+        {
+            return _getTypeStyleId(vectorType->getElementType());
+        }
+        if(auto matrixType = as<IRMatrixType>(type))
+        {
+            return _getTypeStyleId(matrixType->getElementType());
+        }
+        auto style = getTypeStyle(type->getOp());
+        switch (style)
+        {
+        case kIROp_IntType:
+            return 0;
+        case kIROp_FloatType:
+            return 1;
+        case kIROp_BoolType:
+            return 2;
+        case kIROp_PtrType:
+        case kIROp_InOutType:
+        case kIROp_OutType:
+        case kIROp_RawPointerType:
+        case kIROp_RefType:
+            return 3;
+        case kIROp_VoidType:
+            return 4;
+        default:
+            return -1;
+        }
+    }
+
+    IRInst* IRBuilder::emitCast(IRType* type, IRInst* value)
+    {
+        if (isTypeEqual(type, value->getDataType()))
+            return value;
+        
+        auto toStyle = _getTypeStyleId(type);
+        auto fromStyle = _getTypeStyleId(value->getDataType());
+
+        if (fromStyle == kIROp_VoidType)
+        {
+            // We shouldn't be casting from void to other types.
+            SLANG_UNREACHABLE("cast from void type");
+        }
+
+        SLANG_RELEASE_ASSERT(toStyle != -1);
+        SLANG_RELEASE_ASSERT(fromStyle != -1);
+
+        struct OpSeq
+        {
+            IROp op0, op1;
+            OpSeq(IROp op)
+            {
+                op0 = op; op1 = kIROp_Nop;
+            }
+            OpSeq(IROp op, IROp inOp1)
+            {
+                op0 = op; op1 = inOp1;
+            }
+        };
+
+        static const OpSeq opMap[4][5] =
```

Comment:
Some of these conversions seem undesirable. Like a bool to a pointer, or a float to a pointer. Perhaps the point is some earlier checking means this doesn't happen in practice.

Reply:
Right, the type-system already have checks against that. The code here is just being defensive so that it doesn't just crash if the situations occur.